### PR TITLE
Revpi 2647/use serdev highlevel functions aio

### PR DESCRIPTION
**Pull request by linosanfilippo-kunbus**

### CODE REVIEW

**Comment by l1k** [comment]
About commit f1ac37cb7e1c ("piControl: use pibridge_req_io() to send AIO config messages"):

Functionally, the commit is fine. I still have a recommendation: I would amend the definitions of `SAioConfig` and `SAioInConfig` in IoProtocol.h to drop the `uHeader` and `i8uCrc` members.

Next, drop all the code in `piAIOComm_Config()` to populate the `uHeader` fields and calculate the `i8uCrc`. All of that is just dead code because you let `pibridge_req_io()` construct the header and crc. It's confusing to retain that code even though it no longer has any use.

These changes will allow you to define `AIO_CONFIG_DATA1/2/3_LEN` as `sizeof(SAioConfig)` or `sizeof(SAioInConfig)`, respectively, leading to (IMO) clearer code. Alternatively, you could forgo those macro definitions entirely and just use `sizeof()` directly in the `pibridge_req_io()` calls. If you feel that it adds value to document those packet sizes (it probably does), add the information to IoProtocol.h.

**Review by l1k — CHANGES_REQUESTED** [review]

File: piAIOComm.c [hunk 1]
```diff
@@ -266,80 +269,53 @@ INT32U piAIOComm_Config(uint8_t i8uAddress, uint16_t i16uNumEntries, SEntryInfo
...
-			}
-		}
+		config = &aioConfig_s[i];
+		if (config->uHeader.sHeaderTyp1.bitAddress == addr)
+			break;
 	}
 
-	return 4;		// unknown device
+	if (i == i8uConfigured_s)
+		return 4; // unknown device
```
Comment (l1k) [inline piAIOComm.c:290]: Good idea to add this check and thus make it possible to reduce indentation below by 2 tabs.

**Comment by l1k** [comment]
Ugh, I've realized only now that the `aioConfig_s[i].uHeader.sHeaderTyp1.bitAddress` field is necessary to find the `SAioConfig` for a given address. Maybe use a separate `u8[]` array to store the address for a given array index?

File: piAIOComm.c [hunk 2]
```diff
@@ -338,100 +342,44 @@ INT32U piAIOComm_Init(INT8U i8uDevice_p)
...
-	memcpy(last_out[i8uAddress], data_out, sizeof(data_out));
-#endif
-
-	ret = piIoComm_send((INT8U *) & sRequest_l, sizeof(SAioRequest));
-	if (ret == 0) {
-		len_l = sizeof(SAioResponse) - IOPROTOCOL_HEADER_LENGTH - 1;	// data length only
-
-		ret = piIoComm_recv((INT8U *) & sResponse_l, IOPROTOCOL_HEADER_LENGTH + len_l + 1);
-		if (ret > 0) {
-			if (piIoComm_response_valid(&sResponse_l, i8uAddress, len_l)) {
```
[the remark this answers is not in the document]
Comment (linosanfilippo-kunbus) [reply piAIOComm.c:396]: It was by intention to not return an error for this case. The usual API for receiving any kind of (network) messages just requests a buffer that is large enough to store the received data and the buffer length. It usually is not an error if the provided buffer is larger than necessary. I find the (re)interpretation of "buffer length" as "expected length" rather strange. And even though this was the logic implemented in piControl I dont think that we necessarily have to overtake it for the new serdev API...

**Review by l1k — CHANGES_REQUESTED** [review]

File: piAIOComm.c [hunk 3]
```diff
@@ -117,231 +105,194 @@ INT32U piAIOComm_Config(uint8_t i8uAddress, uint16_t i16uNumEntries, SEntryInfo
...
-	return 4;		// unknown device
-}
+	if (dev_idx == num_aios)
+		return 4; // unknown device
+
+	snd_buf = (u8 *) &aioIn1Config_s[dev_idx];
+
+	pr_info_aio("piAIOComm_Init send configIn1\n");
+	ret = pibridge_req_io(addr, IOP_TYP1_CMD_DATA2, snd_buf,
+			      AIO_CONFIG_DATA2_LEN, NULL, 0);
```
[the remark this answers is not in the document]
Comment (linosanfilippo-kunbus) [reply piAIOComm.c:272]: In this function snd_buf is admittedly not as helpful as in piAIOComm_sendCyclicTelegram(). But I would like to keep it, to have a certain consistency across both functions.

File: piAIOComm.c [hunk 4]
```diff
@@ -338,100 +342,45 @@ INT32U piAIOComm_Init(INT8U i8uDevice_p)
...
-#ifdef DEBUG_DEVICE_AIO
-	if (last_out[i8uAddress][0] != sRequest_l.ai8uData[0] || last_out[i8uAddress][1] != sRequest_l.ai8uData[1]) {
-		pr_info_aio("dev %2d: send cyclic Data addr %d output 0x%02x 0x%02x\n",
-			    i8uAddress, RevPiDevice_getDev(i8uDevice_p].i16uOutputOffset,
-			    sRequest_l.ai8uData[0], sRequest_l.ai8uData[1]);
+	ret = pibridge_req_io(addr, IOP_TYP1_CMD_DATA, snd_buf,
+			      AIO_OUTPUT_DATA_LEN, rcv_buf, AIO_INPUT_DATA_LEN);
+	if (ret) {
+		pr_warn_ratelimited("addr %2d: io communication failed with AIO %d\n",
+			addr, ret);
```
[the remark this answers is not in the document]
Comment (linosanfilippo-kunbus) [reply piAIOComm.c:325]: hmm, this is the same indentation that we also use in pibridge.c

**Review by l1k — APPROVED** [review]

**Pull request merged**